R/bmerge.R

```diff
@@ -10,83 +10,239 @@ bmerge <- function(i, x, leftcols, rightcols, io, xo, roll, rollends, nomatch, m
   # This is why shallow() is very importantly internal only, currently.
-
+  
+  # In contrast, x is not passed as shallow, but is shallow copied inside bmerge. 
+  # This allows for coercion of column types for the join, 
+  # but the calling function only sees the original column types of x
+  # careful to only plonk syntax (full column) on x from now on (otherwise x would change)
+  
+  x     = shallow(x)
   origi = shallow(i)      # Needed for factor to factor/character joins, to recover the original levels
                           # Otherwise, types of i join columns are anyways promoted to match x's
                           # types (with warning or verbose)
   resetifactor = NULL     # Keep track of any factor to factor/character join cols (only time we keep orig)
+  
+  ## function to determine type of columns. 
+  ## Important not to use typeof() since integer and factor have same type (integer)
+  ## important to test for is.double instead of is.numeric because of bug with IDate (test 346)
+  getClass <- function(x){
+    if(is.logical(x))                 out <- "logical"
+    else if(is.integer(x))            out <- "integer"
+    else if(base::is.double(x)){ ## base::is.double(integer64) is TRUE, while bit64::is.double(integer64) is FALSE!
+      if(inherits(x, "integer64"))    out <- "integer64" ## is.numeric returns TRUE on integer64 columns
+      ## distinguish integer values with storage mode double from real numeric 
+      ## values that can't be coerced to integer without loss of precision
+      else if(isReallyReal(x))        out <- "reallyDouble"
+      else                            out <- "intAsDouble"
+    }
+    else if(is.character(x))          out <- "character"
+    else if(is.factor(x))             out <- "factor"
+    else                              out <- "other"
+    return(out)
+  }
+  
+  ## function or coercing a vector to a new type. Depending on the source and target type,
+  ## different coercion strategies are used.
+  coerceClass <- function(x, to){
+    sourceType <- getClass(x)
+    ## reallyDouble and intAsDouble are the same here:
+    if(sourceType %chin% c("reallyDouble", "intAsDouble")) sourceType <- "double"
+    if(sourceType == to) return(x) ## nothing to be done
+    if(!to %chin% c("logical", "integer", "double", "character", "factor", "integer64")) 
+      stop("Invalid 'to' argument: ", to)
+    if(sourceType == "other")
+      stop("type coercion not supported for this type: ", paste0(class(x), collapse = ","))
+    if(sourceType %chin% c("logical", "integer", "double") &
+       to         %chin% c("logical", "integer", "double")){
+      ## for these classes, we can do mode coercion to retain other class attributes, e.g. IDate
+      ## identical types have been caught above
+      out       <- x
+      mode(out) <- to
+    }
+    else {
+      ## we need as.'to'() conversion
+      converter <- match.fun(paste0("as.", to))
```

Comment:
there are just few choices here, why not just few `if (...) as.x else if (...) as.y else if ...`? making such converted make perfect sense if we want to support arbitrary coercion.

Reply:
I wanted to be ready for other `as.xxx` conversion, but if you want, I can change.

Reply:
Is there any "other" on roadmap? I am not aware of any. Still it will need to be added to matrix anyway.

Reply:
No there isn't and you are completely riht about the matrix. If you want me to change the implementation of coerceClass, please indicate briefly, I will then do it.